[Python Arrow] Fix issue related to TIMESTAMP_TZ and filter pushdown into PyArrow

Tishj opened the pull request:
This PR fixes #8522 

Arrow supports TIMESTAMP_TZ with any time unit (s, ms, ns, us)
We only support the time unit us for them, so we convert to us.

When a filter is applied that does a comparison against a constant, the constant is cast to the type of the column.
So the constant gets cast to TIMESTAMP_TZ (us)

When this filter gets pushed down into arrow, that results in attempting to compare TIMESTAMP_TZ(ms) against TIMESTAMP_TZ(us), which is not supported.

Tishj committed:
rebind the schema so we can discover what the original arrow type was when creating a scalar

Tishj committed:
add a test for the pyarrow filter pushdown with timestamp_tz values

Tishj committed:
cancel CI

Tishj committed:
Merge remote-tracking branch 'upstream/main' into arrow_timezone_filter

Mytherin commented:
Looks like the exception type is wrong in the test, could you have another look?

Tishj committed:
Merge remote-tracking branch 'upstream/main' into arrow_timezone_filter

Tishj committed:
check the sys modules to see if its loaded don't just rely on 'arrow_dataset().IsLoaded()'

Tishj commented:
Think I found the culprit, the test that's failing is actually underspecified, which I now fixed.
It uses `arrow_dataset().IsLoaded()` which will return true if it was loaded correctly by us before.

The test however monkeypatches the pyarrow dataset to not be present, so in the sys modules it's no longer present.
We don't check for this so the VerifyDatasetLoaded method succeeds, while later on where the attributes are imported this will fail.

I don't see why that's a huge deal though, in both cases it throws an exception, the one we throw is just a little more descriptive, none the less I changed and hopefully fixed it now :)

Tishj commented:
I am tempted to remove the failing test
It's testing a scenario that we can not reliably, deterministically replicate

If the test is run standalone it passes, when I run it together with other tests, it doesn't raise an exception (because pyarrow is loaded before it was monkeypatched out)
And I'm guessing the order tests are run is platform dependent, which makes the exception we get on Linux different because of the sequence of tests that are run before the offending test.

Tishj committed:
remove unreliable undeterministic test

hawkfish commented on tools/pythonpkg/src/python_import_cache.cpp:
```diff
@@ -17,7 +17,11 @@ bool PythonImportCacheItem::LoadSucceeded() const {
 bool PythonImportCacheItem::IsLoaded() const {
 	auto type = (*this)();
-	return type.ptr() != nullptr;
+	bool loaded = type.ptr() != nullptr;
```
Not sure what the point of this change is?

Tishj replied:
None really, I had initially meant to check ModuleIsLoaded if IsLoaded returned true, to double verify
But even with monkeypatching this looked like it didn't make ModuleIsLoaded return false

So this is left ovee code that I might want to undo

Mytherin commented:
Thanks!
